lib/spack/spack/build_environment.py

```diff
@@ -317,7 +317,7 @@ def set_compiler_environment_variables(pkg, env):
     env.set("SPACK_LINKER_ARG", compiler.linker_arg)
 
     # Check whether we want to force RPATH or RUNPATH
-    if spack.config.get("config:shared_linking") == "rpath":
+    if spack.config.get("config:shared_linking:type") == "rpath":
```

Comment:
This still needs to support the old format for backward compatibility.  If you get `config:shared_linking` and it's a string, assume it's the `type`.

We could warn and recommend updating to the new config format; ideally `spack config update` would allow the user to automatically fix it if we're going to warn about it.  I think it's simpler here to just continue accepting a string.  If you want to do the update thing, look at the various `update()` methods in `spack.schema.*` modules for how to handle it.

Reply:
Was already the case